### PR TITLE
Heuristic load balance based on number of particles and number of cells

### PR DESCRIPTION
**Pull request by mrowan137**
This pull request adds the option to load balance according to an heuristic cost based on the number of particles and number of cells on each box, which is computed only at each load balance iteration (resolves issue #713 ).   The option can be enabled in the input file by setting `algo.load_balance_costs_update = heuristic` (the default option is `algo.load_balance_costs_update = timers`).  When the heuristic cost update is enabled, particle and cell weights can be set also in the input file, e.g. `algo.costs_heuristic_cells_wt = 0.1` and `algo.costs_heuristic_particles_wt = 0.9`.

### CODE REVIEW

File: Source/Evolve/WarpXEvolveEM.cpp [hunk 1]
```diff
@@ -779,3 +779,4 @@ WarpX::applyMirrors(Real time){
         }
     }
 }
+
```
Comment (ax3l) [inline Source/Evolve/WarpXEvolveEM.cpp]: Probably has a double EOF newline now. Unnecessary whitespace change:
```suggestion
```

File: Source/WarpX.H [hunk 2]
```diff
@@ -265,6 +274,12 @@ public:
 
     static amrex::Real quantum_xi_c2;
 
+
+    // Load balance options
+    enum LBType { UNDEFINED = -1, DEFAULT, HEURISTIC };
```
Comment (ax3l) [inline Source/WarpX.H]: Can you please add a doxygen string to this enums?
[Here is a small example.](https://github.com/openPMD/openPMD-api/blob/a545854367624984555a1d2a06463e88796df3fd/include/openPMD/IO/AccessType.hpp)

You might also want to make this an `enum struct` or `enum class` in modern C++ :)

File: Source/Parallelization/WarpXRegrid.cpp [hunk 3]
```diff
@@ -11,27 +11,47 @@
 
 using namespace amrex;
 
+std::unique_ptr< Vector<std::unique_ptr<amrex::MultiFab> > > m_costs;
+
 void
-WarpX::LoadBalance ()
+WarpX::LoadBalance (WarpX::LBType load_balance_type/*=DEFAULT*/)
```
Comment (ax3l) [inline Source/Parallelization/WarpXRegrid.cpp]: Verbose, defaults are defined in the header file.
```suggestion
WarpX::LoadBalance (WarpX::LBType load_balance_type)
```

File: Source/WarpX.H [hunk 4]
```diff
@@ -265,6 +274,12 @@ public:
...
+    void LoadBalance (LBType load_balance_type = DEFAULT);
```
Comment (ax3l) [inline Source/WarpX.H]: Please add a Doxygen string to new APIs.
[Here is a small example](https://github.com/openPMD/openPMD-api/blob/a545854367624984555a1d2a06463e88796df3fd/include/openPMD/auxiliary/Date.hpp#L30-L35).

File: Source/WarpX.H [hunk 5]
```diff
@@ -423,6 +438,10 @@ public:
          amrex::IntVect y_nodal_flag, amrex::IntVect z_nodal_flag,
          const int lev);
 
+    // Function to add prtl and cell contributions to costs_heuristic
```
Comment (ax3l) [inline Source/WarpX.H]: Please make this a full-flavored Doxygen string.

File: Source/WarpX.H [hunk 6]
```diff
@@ -423,6 +438,10 @@ public:
...
+    void ComputeCostsHeuristic (int lev);
+    amrex::Real sumMultiFabOverLevs (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& VectorOfMultifabs);
```
Comment (ax3l) [inline Source/WarpX.H]: please add a doxygen string describing what this abbreviated name does and what the args and side-effects of its call are.

File: Source/WarpX.H [hunk 7]
```diff
@@ -653,11 +670,22 @@ private:
...
     amrex::Real const_dt = 0.5e-11;
 
+    // Load balancing
     int load_balance_int = -1;
     amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs;
+    amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs_heuristic;
     int load_balance_with_sfc = 0;
     amrex::Real load_balance_knapsack_factor = 1.24;
 
+    const std::map<std::string, LBType> load_balance_type_str_to_load_balance_type = {
```
Comment (ax3l) [inline Source/WarpX.H]: I would place this map very close to the enum definition, because it might get out of sync on later patches otherwise.

File: Source/WarpX.H [hunk 8]
```diff
@@ -653,11 +670,22 @@ private:
     int n_buffer = 4;
     amrex::Real const_dt = 0.5e-11;
 
+    // Load balancing
     int load_balance_int = -1;
     amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs;
```
Comment (ax3l) [inline Source/WarpX.H]: Please add parameter descriptions `type name; //! what is my unit or how can I find it in AMReX docs` for those members, I fear we will loose track otherwise.

File: Source/WarpX.cpp [hunk 9]
```diff
@@ -1413,3 +1422,4 @@ WarpX::PicsarVersion ()
     return std::string("Unknown");
 #endif
 }
+
```
Comment (ax3l) [inline Source/WarpX.cpp]: Only a single newline at the end of files needed:
```suggestion
```

File: Source/Utils/WarpXUtil.H [hunk 10]
```diff
@@ -128,4 +128,17 @@ T trilinear_interp(T x0, T x1,T y0, T y1, T z0, T z1,
 */
 WarpXParser makeParser (std::string const& parse_function, std::vector<std::string> const& varnames);
 
+namespace {
+// Map with default value (for not-present key)
```
Comment (ax3l) [inline Source/Utils/WarpXUtil.H]: please promote this to a full-flavored doxygen block :) (see example link below)

please include `<map>` in this file.

File: Source/Evolve/WarpXEvolveEM.cpp (hunk 1 above)
Comment (mrowan137) [reply Source/Evolve/WarpXEvolveEM.cpp]: Removed.

File: Source/WarpX.H (hunk 2 above)
Comment (mrowan137) [reply Source/WarpX.H]: Documentation added.  Converted to `enum class` to be modern =)

File: Source/Parallelization/WarpXRegrid.cpp (hunk 3 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Removed.

File: Source/WarpX.H (hunk 4 above)
Comment (mrowan137) [reply Source/WarpX.H]: Documentation added.

File: Source/WarpX.H (hunk 5 above)
Comment (mrowan137) [reply Source/WarpX.H]: Documentation added.

File: Source/WarpX.H (hunk 6 above)
Comment (mrowan137) [reply Source/WarpX.H]: Documentation added.

File: Source/WarpX.H (hunk 7 above)
Comment (mrowan137) [reply Source/WarpX.H]: Moved to after the `enum` definition.

File: Source/WarpX.H (hunk 8 above)
Comment (mrowan137) [reply Source/WarpX.H]: Parameter descriptions added.

File: Source/WarpX.cpp (hunk 9 above)
Comment (mrowan137) [reply Source/WarpX.cpp]: Removed.

File: Source/Utils/WarpXUtil.H (hunk 10 above)
Comment (mrowan137) [reply Source/Utils/WarpXUtil.H]: Documentation added.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 11]
```diff
@@ -228,10 +248,112 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
     }
 }
+
+void
+WarpX::ComputeCostsHeuristic (int lev)
+{
+    MultiFab* cost = WarpX::getCostsHeuristic(lev);
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+    auto nSpecies = mypc.nSpecies();
+    const IntVect& ng = current_fp[lev][0]->nGrowVect();
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]: I do not think we can use that: different `MultiFabs` have different number of guard cells.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 12]
```diff
@@ -228,10 +248,112 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+        Array4<Real> const& costarr = cost->array(mfi);
+        amrex::ParallelFor(
+                           tbx,
+                           [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                           {
+                               costarr(i,j,k) += wt;
+                           }
+                          );
+        }
+    }
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]: I think this could be replaced by
```c++
    int n_cells = 0;
    Multifab* Ex = Efield_fp[lev][0].get();
    for (MFIter mfi(*Ex, false); mfi.isValid(); ++mfi){
        const Box& gbx = mfi.growntilebox();
        n_cells += gbx.numPts();
    }
```
Then, we could do
```c++
    cost[lev] = costs_heuristic_particles_wt * n_particles +
                costs_heuristic_cells_wt * n_cells;
```

File: Source/Parallelization/WarpXRegrid.cpp [hunk 13]
```diff
@@ -11,27 +11,47 @@
...
+    switch (load_balance_type)
+    {
+    case LBType::DEFAULT:
+        m_costs.reset(&costs);
+        break;
+    case LBType::HEURISTIC:
+        for (int lev = 0; lev <= finest_level; ++lev)
+        {
+            WarpX::ComputeCostsHeuristic(lev);
+        }
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Loop moved inside.

File: Source/WarpX.H [hunk 14]
```diff
@@ -265,6 +274,31 @@ public:
...
+        HEURISTIC       /**< load balance according to weights computed from number of cells
+                             and number of particles per box (i.e., with `costs_heuristic`)*/
+    }; // LBType
+
+    /** Map from string to `LBType`; used in parsing input.
+     */
+    const std::map<std::string, LBType> load_balance_type_str_to_load_balance_type = {
+        {"default",   LBType::DEFAULT},
+        {"heuristic", LBType::HEURISTIC}
+    };
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/WarpX.H]: Switched to the procedure in `WarpXAlgorithmSelection`.  Now using `timers` and `heuristic.`

File: Source/WarpX.H [hunk 15]
```diff
@@ -265,6 +274,31 @@ public:
...
     static amrex::Real quantum_xi_c2;
 
+
+    /** Strategy to compute weights for use in load balance.
+     */
+    enum class LBType
+    {
+        UNDEFINED = -1, //!< default load balance type if invalid `warpx.load_balance_type` in input
+        DEFAULT,        //!< load balance according to in-code timer-based weights (i.e., with  `costs`)
+        HEURISTIC       /**< load balance according to weights computed from number of cells
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/WarpX.H]: Done.

File: Source/WarpX.cpp [hunk 16]
```diff
@@ -1037,6 +1045,7 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
 
     if (load_balance_int > 0) {
         costs[lev].reset(new MultiFab(ba, dm, 1, 0));
+        costs_heuristic[lev].reset(new MultiFab(ba, dm, 1, 0));
     }
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/WarpX.cpp]: I have added this logic here (and other places as necessary; see most recent changes).  A related point: `costs` is not computed if `heuristic` load balance cost update is used, and so also will not be output with plotfiles.

File: Source/WarpX.H [hunk 17]
```diff
@@ -653,10 +699,30 @@ private:
...
+                                                          assigned to a rank during load balance via the
+                                                          'knapsack' strategy; e.g., if there are 4 boxes
+                                                          per rank, `load_balance_knapsack_factor=2` limits
+                                                          the maximum number of boxes that can be assigned
+                                                          to a rank to 8. */
+    std::string load_balance_type_str = "default"; /**< Method, read from input, for computing load balance
+                                                        weights (e.g. `costs`, `costs_heuristic`); input
+                                                        selection is converted to `LBType` for internal use. */
+    LBType load_balance_type = LBType::DEFAULT; /**< Method for computing load balance weights (e.g. `costs`,
+                                                     `costs_heuristic`). */
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/WarpX.H]: Removed.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 18]
```diff
@@ -228,10 +248,112 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+                Real wt = (costs_heuristic_prtl_wt*(1.0*np)/tbx.d_numPts()); // prtl wt per (work) cell
+                Array4<Real> const& costarr = cost->array(pti);
+                amrex::ParallelFor(
+                                   tbx,
+                                   [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                                   {
+                                       costarr(i,j,k) += wt;
+                                   }
+                                  );
+            }
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Replaced.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 19]
```diff
@@ -228,10 +248,112 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+                        total += mfArray(i, j, k);
+                    }
+                }
+            }
+        }
+    }
+
+    ParallelDescriptor::ReduceRealSum(total);
+    return total;
+}
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Removed.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 11 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Changed to match the suggestion for 'comment on lines 298 to 334'.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 12 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Replaced.  As a double check, I confirm that I compute identical distribution mappings whether using (now) vector of reals, or the (old) calculation with MultiFabs.

File: Source/WarpX.H [hunk 20]
```diff
@@ -653,10 +699,30 @@ private:
...
-    amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs;
-    int load_balance_with_sfc = 0;
-    amrex::Real load_balance_knapsack_factor = 1.24;
+    // Load balancing
+    int load_balance_int = -1; /**< Interval to perform load balance; `load_balance_int=5`, e.g., will
+                                    load balance during steps 5, 10, 15, etc. */
+    amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs; /**< Collection of multifabs to keep track
+                                                                 of weights, based on in-code timers,
+                                                                 for use in load balancing routines. */
+    amrex::Vector<std::unique_ptr<amrex::MultiFab> > costs_heuristic; /**< Collection of multifabs to keep track
```
[the remark this answers is not in the document]
Comment (mrowan137) [reply Source/WarpX.H]: Agree.  It is updated to use a vector of reals instead of MultiFab.

File: Source/Evolve/WarpXEvolveEM.cpp [hunk 21]
```diff
@@ -61,25 +61,25 @@ WarpX::EvolveEM (int numsteps)
...
             amrex::Abort("LoadBalance for PSATD: TODO");
 #endif
-
             if (step > 0 && (step+1) % load_balance_int == 0)
             {
                 LoadBalance();
-                // Reset the costs to 0
-                for (int lev = 0; lev <= finest_level; ++lev) {
-                    costs[lev]->setVal(0.0);
-                }
```
Comment (mrowan137) [inline Source/Evolve/WarpXEvolveEM.cpp]: @MaxThevenet I have removed this because `RemakeLevel` called in `LoadBalance` seems to do the same thing.  Is this really redundant or does it guard against a special case?  Please see `RemakeLevel` in WarpXRegrid.cpp..

Comment (MaxThevenet) [reply Source/Evolve/WarpXEvolveEM.cpp]: Thanks for pointing this out. It seems that you're right, but there could be some corner case for which this safeguard is useful. Maybe @WeiqunZhang or @atmyers would know. For safety, I would let it as it was, as it is not directly the topic of this PR.

File: Source/Evolve/WarpXEvolveEM.cpp [hunk 22]
```diff
@@ -92,13 +92,13 @@ WarpX::EvolveEM (int numsteps)
             FillBoundaryB(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
             UpdateAuxilaryData();
             // on first step, push p by -0.5*dt
-            for (int lev = 0; lev <= finest_level; ++lev) {
+            for (int lev = 0; lev <= finest_level; ++lev)
+            {
```
Comment (MaxThevenet) [inline Source/Evolve/WarpXEvolveEM.cpp]: In general, could you avoid these style changes in a non-style-related PR? They dilute a little bit the topic of this PR (no need to remove them from this PR as there are not too many of these changes). Thanks!

File: Source/WarpX.H [hunk 23]
```diff
@@ -425,6 +450,12 @@ public:
          amrex::IntVect y_nodal_flag, amrex::IntVect z_nodal_flag,
          const int lev);
 
+    /** \brief add particle and cell contributions to multifabs in `costs_heuristic`
+     *
+     * \param lev level for which to add particle and cell contributions to `costs_heuristic`
```
Comment (MaxThevenet) [inline Source/WarpX.H]:
```suggestion
```

File: Source/WarpX.H [hunk 24]
```diff
@@ -425,6 +450,12 @@ public:
          amrex::IntVect y_nodal_flag, amrex::IntVect z_nodal_flag,
          const int lev);
 
+    /** \brief add particle and cell contributions to multifabs in `costs_heuristic`
+     *
```
Comment (MaxThevenet) [inline Source/WarpX.H]:
```suggestion
```

File: Source/WarpX.H [hunk 25]
```diff
@@ -655,10 +684,27 @@ private:
...
     int load_balance_with_sfc = 0;
+    /** Controls the maximum number of boxes that can be assigned to a rank during
+     * load balance via the 'knapsack' strategy; e.g., if there are 4 boxes per rank,
+     * `load_balance_knapsack_factor=2` limits the maximum number of boxes that can
+     * be assigned to a rank to 8. */
     amrex::Real load_balance_knapsack_factor = 1.24;
+    /** Weight factor for cells in updating `costs_heuristic`. */
+    amrex::Real costs_heuristic_cells_wt = 0.1;
+    /** Weight factor for particles in updating `costs_heuristic`. */
+    amrex::Real costs_heuristic_particles_wt = 0.9;
```
Comment (MaxThevenet) [inline Source/WarpX.H]: Thanks for these comments!

File: Source/Parallelization/WarpXRegrid.cpp [hunk 26]
```diff
@@ -30,7 +43,49 @@ WarpX::LoadBalance ()
...
+        amrex::Real efficiency;
+        const bool do_full_knapsack = true;
+        const int nmax = static_cast<int>(std::ceil(nboxes/nprocs*load_balance_knapsack_factor));
+
+        newdm.KnapSackProcessorMap(cost_heuristic, nprocs, &efficiency, do_full_knapsack, nmax);
+
+        RemakeLevel(lev, t_new[lev], boxArray(lev), newdm);
+    }
     mypc->Redistribute();
 }
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]: For knapsack, could we simply call
```c++
static DistributionMapping makeKnapSack   (const Vector<Real>& rcost);
```
so that `WarpX:: LoadBalanceHeuristic` is almost identical to `WarpX::LoadBalanceTimers`?

And, @atmyers @WeiqunZhang, could `makeSFC` have the option to take a vector of Real as an input argument, consistent with `makeKnapSack`?

File: Source/Parallelization/WarpXRegrid.cpp [hunk 27]
```diff
@@ -223,15 +276,61 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
         amrex::Abort("RemakeLevel: to be implemented");
     }
 }
+
+void
+WarpX::ComputeCostsHeuristic ()
+{
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        long n_particles = 0;
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]:
```suggestion
```
If you're not using this variable to accumulate the number of particles, I don't think it is useful anymore.

File: Source/Parallelization/WarpXRegrid.cpp [hunk 28]
```diff
@@ -223,15 +276,61 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+        long n_cells = 0;
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]:
```suggestion
```

File: Source/Parallelization/WarpXRegrid.cpp [hunk 29]
```diff
@@ -223,15 +276,61 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+
+        // Species loop
+        for (int i_s = 0; i_s < nSpecies; ++i_s)
+        {
+            auto & myspc = mypc.GetParticleContainer(i_s);
+
+            // Particle loop
+            for (WarpXParIter pti(myspc, lev); pti.isValid(); ++pti)
+            {
+                n_particles = pti.numParticles();
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]:
```suggestion
```

File: Source/Parallelization/WarpXRegrid.cpp [hunk 30]
```diff
@@ -223,15 +276,61 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
...
+                (*costs_heuristic[lev])[pti.index()] += costs_heuristic_particles_wt*n_particles;
```
Comment (MaxThevenet) [inline Source/Parallelization/WarpXRegrid.cpp]:
```suggestion
                (*costs_heuristic[lev])[pti.index()] += costs_heuristic_particles_wt*pti.numParticles();
```
Would this work? And similar for the number of cells.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 26 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: If we call this way, `load_balance_knapsack_factor` will have no effect when using the heuristic costs update (the way it is currently written allows `load_balance_knapsack_factor` to work the same way whether using `timers` or `heuristic`).  Ideally there would be a call like `static DistributionMapping makeKnapSack   (const Vector<Real>& rcost, const int nmax)` in AMReX.  Should we leave this as is or use `DistributionMapping makeKnapSack   (const Vector<Real>& rcost)` (and add documentation that `load_balance_knapsack_factor` does not work with heuristic costs update)?

File: Source/Evolve/WarpXEvolveEM.cpp (hunk 22 above)
Comment (mrowan137) [reply Source/Evolve/WarpXEvolveEM.cpp]: Yes, will avoid this in the future!

File: Source/Parallelization/WarpXRegrid.cpp (hunk 26 above)
Comment (MaxThevenet) [reply Source/Parallelization/WarpXRegrid.cpp]: @atmyers is working on implementing
```c++
DistributionMapping::makeSFC(*costs[lev], false)
DistributionMapping::makeKnapSack(*costs[lev], nmax);
```
where `costs[lev]` is an `amrex::Vector<amrex::Real>`. Will that work?

File: Source/Evolve/WarpXEvolveEM.cpp (hunk 21 above)
Comment (mrowan137) [reply Source/Evolve/WarpXEvolveEM.cpp]: I've reverted to clearing levels here.

**Comment by atmyers** [comment]
We added new versions of `makeKnapSack` and `makeSFC` that my be of use here: https://github.com/AMReX-Codes/amrex/pull/709/files

File: Source/Parallelization/WarpXRegrid.cpp (hunk 26 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Yes sounds good, I will add it when the call is available.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 27 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Removed.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 28 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Removed.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 29 above)
Comment (mrowan137) [reply Source/Parallelization/WarpXRegrid.cpp]: Switched to this.

File: Source/WarpX.H [hunk 31]
```diff
@@ -199,8 +202,24 @@ public:
     std::vector<bool> getPMLdirections() const;
 
     static amrex::MultiFab* getCosts (int lev) {
-        if (m_instance) {
+        if (WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers) {
+            if (m_instance) {
             return m_instance->costs[lev].get();
+        } else {
+            return nullptr;
```
Comment (MaxThevenet) [inline Source/WarpX.H]:
```suggestion
            } else {
                return nullptr;
```
